corehq/apps/couch_sql_migration/couchsqlmigration.py

```diff
@@ -742,6 +775,22 @@ def get_queue_obj_id(queue_obj):
             return getattr(queue_obj, queue_id_param)
         self.get_queue_obj_id = get_queue_obj_id
 
+        def get_cached_list_key():
+            return "partial_queues.queued_or_processing.%s" % run_timestamp
+        self.get_cached_list_key = get_cached_list_key
+
+    def add_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        client.rpush(self.get_cached_list_key(), doc_id)
+
+    def remove_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        gevent.spawn(client.lrem, self.get_cached_list_key(), 1, doc_id)
```

Comment:
From https://github.com/dimagi/commcare-hq/pull/21007#discussion_r197841021

> i could spin it off into a greenlet since this could be a bottleneck and, since redis is single threaded before v4 (which i don't think we're on), and is supposed to be thread safe, it wouldn't be an issue

Doing this in a greenlet means that other processing can continue, possibly allowing more items to be added to the queue before `lrem` in this greenlet gets invoked. However, since `lrem` is atomic it _will_ block other mutations of that key at some point, which means that this could still be a bottleneck. We just have less insight into when it will happen because the order of operations now depends on the gevent task scheduler. Do you think there is enough other processing not involving this redis key that this will benefit from concurrency? 

Aside: you may already know this, but it is not trivial to implement a durable task queue with redis (see this [SO question](https://stackoverflow.com/q/27986649/10840) and the [article in the first comment](http://blog.bronto.com/engineering/reliable-queueing-in-redis-part-1/))—it takes more than one list to do it properly. Is that what we're trying to accomplish here?

Reply:
Yeah, there's a decent bit of other processing time because of the numerous requests to couch and pg.

A few notes on my thinking about the redis operation: 
(1) Redis tends to be pretty snappy even for O(n) ops (I did some benchmarking if you're interested, but summary: for a list with 100000 objects, trying to remove something that doesn't exist (the worst case), takes <=4ms always and mostly ~1ms, so network latency might be the main blocker anyway). 
(2) n is rarely going to be that large so this would not be a bottleneck in the vast majority of circumstances. 
(3) And because this isn't an academic question, I'm going to make the argument that even where it is large, this will be removing objects from the front of the list primarily, so worst case scenario analysis isn't the most necessary.
Also, because the other mutations are also I/O operations, it won't block the entire process and I think the pool should be big enough to handle some blocking. 
Last thing was that this is the final operation of the greenlets in the pool, so I think it's better to open up space for a new greenlet rather than block here.
So basically, at the very least it isn't worse to have it scheduled by gevent.

Reply:
To the aside --> No. This is meant to be a best guess that has _at least_ all the ids that need to be processed. If it reprocesses one, it will error out in pg and continue on rather than saving, with no harm done to the process except an extra ~0.1 seconds.

However, that article was an interesting read, at the least because it describes something not too far off from what I implemented without redis

Reply:
Excellent! Sounds like you've thought through this pretty thoroughly. Thanks for the detailed responses to my pessimistic questions.